single_offer/src/lib.rs

```diff
@@ -13,202 +12,146 @@ mod token {
+    pub fn create(
         e: Env,
-        admin: Identifier,
+        seller: Address,
         sell_token: BytesN<32>,
         buy_token: BytesN<32>,
-        n: u32,
-        d: u32,
+        sell_price: u32,
+        buy_price: u32,
     ) {
-        if has_administrator(&e) {
-            panic!("admin is already set");
+        if e.storage().has(&DataKey::Offer) {
+            panic!("offer is already created");
         }
-
-        if d == 0 {
-            panic!("d is zero but cannot be zero");
+        if buy_price == 0 || sell_price == 0 {
+            panic!("zero price is not allowed");
         }
-
-        write_administrator(&e, admin);
-
-        put_sell_token(&e, sell_token);
-        put_buy_token(&e, buy_token);
-        put_price(&e, Price { n, d });
+        // Authorize the `create` call by seller to verify their identity.
+        seller.require_auth();
+        write_offer(
+            &e,
+            &Offer {
+                seller,
+                sell_token,
+                buy_token,
+                sell_price,
+                buy_price,
+            },
+        );
     }
 
-    fn trade(e: Env, to: Identifier, min: i128) {
-        let balance_buy_token = get_balance_buy(&e);
-
-        let price = load_price(&e);
-
-        let amount = balance_buy_token * price.d as i128 / price.n as i128;
-
-        if amount < min {
-            panic!("will receive less than min");
-        }
-
-        transfer_sell(&e, to, amount);
-
-        let admin = read_administrator(&e);
-        transfer_buy(&e, admin, balance_buy_token);
+    // Trades `buy_token_amount` of buy_token from buyer for `sell_token` amount
+    // defined by the price.
+    // Buyer needs to authorize the `trade` call and internal `xfer` call to
+    // the contract address.
+    pub fn trade(e: Env, buyer: Address, buy_token_amount: i128) {
+        // Buyer needs to authorize the trade.
+        buyer.require_auth();
+
+        // Load the offer and prepare the token clients to do the trade.
+        let offer = load_offer(&e);
+        let sell_token_client = token::Client::new(&e, &offer.sell_token);
+        let buy_token_client = token::Client::new(&e, &offer.buy_token);
+
+        // Compute the amount of token that buyer needs to receive.
+        let sell_token_amount = buy_token_amount
+            .checked_mul(offer.sell_price as i128)
+            .unwrap_optimized()
+            / offer.buy_price as i128;
+
+        let contract = e.current_contract_address();
+
+        // Perform the trade in 3 `xfer` steps.
+        // Note, that we don't need to verify any balances - the contract would
+        // just trap and roll back in case if any of the transfers fails for
+        // any reason, including insufficient balance.
+
+        // Transfer the `buy_token` from buyer to this contract.
+        // This `xfer` call should be authorized by buyer.
+        // This could as well be a direct transfer to the seller, but sending to
```

Comment:
Should we encourage this? This contract would be cheaper to use with one less xfer.

Reply:
That's true, but in a general case it's probably hard to achieve (and I really believe that limiting the scope of trust would have a serious UX benefit). For the next release we can address this with something like `xfer_contract` we've discussed at some point, which basically would fuse allowance+xfer_from into a single call (sign the max_amount+contract, then use actual amount/actual recipient at execution time).

Reply:
Yeah that's a valid point. I was thinking that that the call stack should be sufficient for trust, but it still may not be from the users perspective